# MIG-Phylogenomics
## 0. [Dependencies](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Dependencies.ipynb)
## 1. [CDSs and proteins from genome assemblies](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/CDSs_and_proteins_from_genome_assemblies.ipynb)
## 2. [Map-assemble genes from read data for samples without assemblies](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Map_assemble_genes.ipynb)
## 3. [Orthology clustering](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Orthology_clustering.ipynb)
## 4. [Nuclear phylogenomics](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Nuclear_phylogenomics.ipynb)
## 5. [Mitochondrial genome assembly](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Mitochondrial_genome_assembly.ipynb)
## 6. [Mitochondrial genome annotation](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Mitochondrial_genomes_annotation.ipynb)
## 7. [Mitochondrial genome phylogenomics](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Mitochondrial_genomes_tree.ipynb)
## 8. [Intra-genome identity among homeologue gene pairs](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Intra_genome_sequence_divergence.ipynb)
## 9. [Coverage ratio between homeologue contigs within a genome](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Median_ratio.ipynb)
## 10. [Gene conversion in *M. floridensis* MfloSJF1](https://github.com/HullUni-bioinformatics/MIG-Phylogenomics/blob/master/Non-conversion_tracts_in_MfloSJF1.ipynb)